### PR TITLE
fix(diagnostics): translate legacy signs before diagnostic scheduler

### PR DESCRIPTION
**Pull request by kawre**
Changed the location of the legacy diagnostic sign handler.
By doing so the `vim.diagnostic.config().signs` will get converted no matter the `update_in_insert` value.

This will resolve some unexpected behavior in plugins like `neotree` because if `update_in_insert` is set to `false` and once a lsp loads and you are in `insert mode` the signs will not be yet converted but the plugin will get notified of the published diagnostics and expect the `signs.text` to be defined.

https://github.com/nvim-neo-tree/neo-tree.nvim/assets/69250723/706a396d-51d6-47cf-ae47-21eed9dba4c5

more info there
https://github.com/nvim-neo-tree/neo-tree.nvim/issues/1290

### CODE REVIEW

**Comment by kawre** [comment]
@gpanders Just to make sure we are on the same page, I'm posting a video with the `neotree` fix included.

https://github.com/neovim/neovim/assets/69250723/bb82da45-78a3-4dc4-8c7c-4f783f302e58

**Review by justinmk — APPROVED** [review]
Seems fine to me. Can you rebase?

 cc @gpanders 

ref #26618

**Comment by gpanders** [comment]
I mentioned in https://github.com/nvim-neo-tree/neo-tree.nvim/issues/1290 why I don’t think this is the right solution (or even necessary). The neo-tree issue is closed, does this PR still actually solve any problem?

**Comment by kawre** [comment]
I believe there is a misunderstanding. @gpanders you only explained that `vim.diagnostic.config().signs.text` can be `nil`, yet this PR doesn't tackle that.
 
This PR is just ensuring that diagnostic signs are translated before `diagnostic scheduler` because as seen in the video I posted before, you can see that if you boot neovim and don't leave insert mode with `update_in_insert` disabled, diagnostic signs will not get translated until you leave insert mode and only then they will update.

File: runtime/lua/vim/diagnostic.lua [hunk 1]
```diff
@@ -908,48 +953,7 @@ M.handlers.signs = {
...
-            opts.signs.numhl[severity] = sign.numhl
-          end
-
-          if opts.signs.linehl[severity] == nil then
-            opts.signs.linehl[severity] = sign.linehl
-          end
-        end
-      end
-    end
+    resolve_legacy_diagnostic_sign_definitions(opts)
```
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:956]: Do we still need this here? Is the call in `vim.diagnostic.show()` not sufficient?

File: runtime/lua/vim/diagnostic.lua [hunk 2]
```diff
@@ -1283,6 +1287,8 @@ function M.show(namespace, bufnr, diagnostics, opts)
 
   opts = get_resolved_options(opts, namespace, bufnr)
 
+  resolve_legacy_diagnostic_sign_definitions(opts)
```
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:1290]:
```suggestion
  -- TODO(#26927): Remove this in 0.12 release cycle
  resolve_legacy_diagnostic_sign_definitions(opts)
```

File: runtime/lua/vim/diagnostic.lua (hunk 1 above)
Comment (gpanders) [reply runtime/lua/vim/diagnostic.lua]: @kawre Did you verify that this still resolves your issue even when removing the other call site?

**Pull request closed without merging**